File: relay/src/main/java/net/betrayd/webspeak/relay/ServerPlayerConnection.java
package net.betrayd.webspeak.relay;

import java.util.Map;

import org.eclipse.jetty.websocket.api.Callback;
import org.eclipse.jetty.websocket.api.Session;
import org.eclipse.jetty.websocket.api.StatusCode;
import org.eclipse.jetty.websocket.api.annotations.OnWebSocketClose;
import org.eclipse.jetty.websocket.api.annotations.OnWebSocketMessage;
import org.eclipse.jetty.websocket.api.annotations.OnWebSocketOpen;
import org.eclipse.jetty.websocket.api.annotations.WebSocket;
import org.slf4j.LoggerFactory;

import net.betrayd.webspeak.relay.ServerConnection.LinkedConnection;

/**
 * The server-bound side of a player connection.
 */
@WebSocket
public class ServerPlayerConnection {

    private Session session;
    ServerConnection server;
    String sessionID;

    // store this here to handle a lot of packets fast without cpu doing map stuff.
    private ClientPlayerConnection connectedClient = null;

    /**
     * Called when the client-bound side has connected.
     * @param connectionAddress The address of the connected client.
     * @param connection The client connection.
     */
    public void connectedClient(String connectionAddress, ClientPlayerConnection connection) {
        this.connectedClient = connection;
        // We only use json once (here), so no reason to setup a whole serialization system
        this.send("relayClientConnect;{context:\"" + connectionAddress + "\"}");
    }

    /**
     * Called when the client-bound side has disconnected.
     * @param statusCode Disconnect status code.
     * @param reason Disconnect reason.
     */
    public void disconnectClient(int statusCode, String reason) {
        this.connectedClient = null;
        this.send("relayClientDisconnect;{statusCode:" + statusCode + ",reason:\"" + reason + "\"}");
    }

    /**
     * Close the connection and associated client connection.
     */
    public void disconnect() {
        session.close();
    }

    public void send(String message) {
        session.sendText(message, Callback.NOOP);
    }

    @OnWebSocketOpen
    public synchronized void onWebSocketOpen(Session session) {
        LoggerFactory.getLogger(getClass()).info("Server opened player with params: {}, server: {}",
                session.getUpgradeRequest().getQueryString(), session.getRemoteSocketAddress());
        
        this.session = session;
        
        Map<String, String> query = NetUtils.splitQueryString(session.getUpgradeRequest().getQueryString());
        String privateServerID = query.get("key");
        String publicServerID = query.get("server");

        server = WebSpeakRelay.servers.get(publicServerID);
        if (server != null && server.usesPrivateKey(privateServerID)) {
            sessionID = query.get("id");
            if (server.connections.putIfAbsent(sessionID, new LinkedConnection(this)) != null) {
                session.close(StatusCode.BAD_DATA, "Session already exists with that ID.", Callback.NOOP);
                return;
            };
        } else {
            session.close(StatusCode.BAD_DATA, "Invalid server/key pair", Callback.NOOP);
        }
    }

    // I don't know about this implementation, as the server is a client so when it
    // closes it's not going to say what we want but EH I don't actually care
    @OnWebSocketClose
    public synchronized void OnWebSocketClose(int statusCode, String reason) {
        LoggerFactory.getLogger(getClass()).info("Server player connection closed. PlayerID: {}, Server: {}", sessionID,
                session.getRemoteSocketAddress());
        
        if (this.connectedClient != null) {
            this.connectedClient.close(statusCode, reason);
            this.connectedClient = null;
        }
        // ClientPlayerConnection client;
        // synchronized(this) {
        //     client = this.connectedClient;
        //     this.connectedClient = null;
        // }

        // if (client != null) {
        //     client.close(statusCode, reason);
        // }
        server.connections.remove(sessionID);
    }

    @OnWebSocketMessage
    public void onWebSocketMessage(String message) {
        // Store for thread safety
        ClientPlayerConnection client = connectedClient;
        if (client == null) {
            return;
        }
        client.send(message);
    }
}

File: serverPlayerConnection.ts
import { ClientPlayerConnection } from "./clientPlayerConnection.ts";
import { ServerConnection } from "./serverConnection.ts";
import { LinkedConnection } from "./linkedConnection.ts";
import { servers } from "./index.ts";

export class ServerPlayerConnection {
    socket!: WebSocket;
    requestURL!: string;
    server!: ServerConnection | undefined;
    sessionID: string | null = null;
    private connectedClient: ClientPlayerConnection | null = null;
    constructor(socket: WebSocket, requestURL: string) {
        socket.addEventListener("open", this);
        socket.addEventListener("message", this);
        socket.addEventListener("close", this);
        this.requestURL = requestURL;
        this.socket = socket;
    }

    handleEvent(event: Event) 
    {
        switch(event.type)
        {
            case "open":
                this.eventOpen();
                break;
            case "message":
                if(event instanceof MessageEvent)
                    {
                        if(typeof event.data != "string")
                        {
                            Error("we were sent non string data in message!");
                            return;
                        }
                        this.eventMessage(event);
                    }
                break;
            case "close":
                if(event instanceof CloseEvent)
                {
                    this.eventClose(event);
                }
                break;
        }
    }

    public connectClient(connectionAddress: string, connection: ClientPlayerConnection) {
        this.connectedClient = connection;
        this.socket.send("relayClientConnect;{context:\"" + connectionAddress + "\"}");
    }

    public disconnectClient(statusCode: number, reason: string) {
        this.connectedClient = null;
        this.socket.send("relayClientDisconnect;{statusCode:" + statusCode + ",reason:\"" + reason + "\"}");
    }

    public disconnect(code?: number, reason?: string) {
        this.socket.close(code, reason);
    }

    public send(message: string) {
        this.socket.send(message);
    }

    private eventOpen() {
        const params: URLSearchParams = new URL(this.requestURL).searchParams;
        const privateServerID = params.get("key");
        const publicServerID = params.get("server");
        this.sessionID = params.get("id");

        console.log("Server opened player with id: ", this.sessionID, "server:", publicServerID);

        if (publicServerID == null || privateServerID == null || this.sessionID == null) {
            this.disconnect(1002, "Not correct parameters");
            console.error("Not correct parameters");
            return;
        }
        this.server = servers.get(publicServerID);
        if (this.server != undefined && this.server.usesPrivateKey(privateServerID)) {
            if (this.server.connections.has(this.sessionID)) {
                this.disconnect(1002, "Session already exists with that ID.");
                return
            }
            this.server.connections.set(this.sessionID, new LinkedConnection(this));
        }
        else {
            this.disconnect(1002, "Invalid server/key pair");
        }

    }

    private eventMessage(event: MessageEvent<string>) {
        const client: ClientPlayerConnection | null = this.connectedClient;
        if (client == null) {
            return;
        }
        client.send(event.data);
    }

    private eventClose(event: CloseEvent) {
        if (this.connectedClient != null) {
            this.connectedClient.disconnect(event.code, event.reason);
            this.connectedClient = null;
        }
        if (this.server != undefined && this.sessionID != null) {
            this.server.connections.delete(this.sessionID);
        }
    }

}